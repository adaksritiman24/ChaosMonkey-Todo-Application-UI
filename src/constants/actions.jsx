export const ACTIONS = {
    CHANGE_SCOPE : "change scope",
    SET_IS_CREATE_TODO_MODAL_OPEN : "set is create todo modal open",
    SET_TODOS : "set todos",
}
