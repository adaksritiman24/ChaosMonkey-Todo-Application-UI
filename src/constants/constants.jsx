export const SCOPE = {
    ALL : "all",
    COMPLETED : "completed",
    INCOMPLETE : "incomplete",
}

export const state = {
    scope : SCOPE.ALL,
    isCreateTodoModalOpen : false,
    loading : true,
}

export const baseURI = "http://localhost:8080/todoservice/v1";