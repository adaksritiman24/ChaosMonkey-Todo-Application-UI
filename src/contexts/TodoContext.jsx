import { createContext } from "react";

const initialContextValue = {
    state : null,
    dispatch : ()=> {}
}

const TodoContext = createContext(initialContextValue);

export default TodoContext;