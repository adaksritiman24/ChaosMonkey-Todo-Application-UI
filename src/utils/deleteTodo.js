import axios from "axios";
import { baseURI } from "../constants/constants"

export const deleteTodo = async(id)=> {
    const config = {
        method : "delete",
        url : baseURI + `/todo/${id}`,
        headers : {} 
    };

    const {data} = await axios(config);
    return data.todoId;
}
