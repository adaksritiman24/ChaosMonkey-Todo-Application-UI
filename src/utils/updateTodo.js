import axios from "axios";
import { baseURI } from "../constants/constants";

const updateTodo= async(newTodo)=>{
    const todoData = JSON.stringify({
        "title": newTodo.title,
        "body": newTodo.body,
        "completed": newTodo.completed,
      });

     const config = {
        method: 'patch',
        url: baseURI+'/todo/'+newTodo.id,
        headers: { 
          'Content-Type': 'application/json'
        },
        data : todoData,
      };

    const {data} =await axios(config);
    return data;
}

export default updateTodo;