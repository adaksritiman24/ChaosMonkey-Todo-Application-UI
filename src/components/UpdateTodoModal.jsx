import { Delete } from '@mui/icons-material'
import { Box, Button, FormControl, IconButton, Modal, Stack, styled, TextField, Tooltip, Typography, useTheme } from '@mui/material'
import { grey } from '@mui/material/colors'
import React, { useContext } from 'react'
import { ACTIONS } from '../constants/actions'
import TodoContext from '../contexts/TodoContext'
import updateTodo from '../utils/updateTodo'
import deleteTodo from '../utils/deleteTodo'

const StyledTextField = styled(TextField)({
    fontFamily : "helvetica",
    
    "& .MuiInputBase-input" : {
        fontSize : "24px",
    }
})

const UpdateTodoModal=()=> {

    const {appState, dispatch} = useContext(TodoContext);

    const todo = appState.updatingTodo;
    if(!todo) return <></>;

    const theme = useTheme();
    const tabletView = theme.breakpoints.up("md");

    const handleModalClose = ()=> {
        dispatch({
            type : ACTIONS.SET_IS_UPDATE_TODO_MODAL_OPEN,
            payload : false,
        })
        dispatch({
            type : ACTIONS.SET_UPDATING_TODO,
            payload : undefined,
        })
    }


    const handleSubmit = async (e)=> {
        e.preventDefault();
        const {title, description} = e.target; 
        const cleanedTitle = title.value.trim();
        const cleanedBody = description.value.trim();

        const changedTodo = {...todo, title: cleanedTitle, body: cleanedBody};
        const updatedTodo = await updateTodo(changedTodo);

        const updatedAllTodos = appState.todos.map((todo)=>{
            if(todo.id === updatedTodo.id)
                return updatedTodo;
            return todo;    
        })

        dispatch({
            type : ACTIONS.SET_TODOS,
            payload : updatedAllTodos
        })

        handleModalClose();
    }

    const handleDeleteTodo = async()=> {
        const todoId = await deleteTodo(todo.id);

        const updatedAllTodos = appState.todos.filter((todo)=>todo.id !== todoId)
        dispatch({
            type : ACTIONS.SET_TODOS,
            payload : updatedAllTodos
        })
        handleModalClose();
    }

  return (
    <Modal
        open={appState.isUpdateTodoModalOpen}
        onClose={handleModalClose}
        aria-labelledby="update-modal-modal-title"
        aria-describedby="update-modal-modal-description"
        data-testid = "update-todo-modal"
    >
        <Box sx={{
            position : 'absolute',
            background: grey[100],
            transform : "translate(-50%, -50%)",
            left: "50%",
            top: "35%",
            width : "270px",
            [tabletView] : {
                width : "400px",
            },
            p : 3,
            borderRadius:1,
        }}>
            <Typography id="update-modal-modal-title" variant="h4" component="h2">
               Update Todo
            </Typography>
            <hr/>
            <form onSubmit={handleSubmit}>
                <Stack spacing={2}>
                    <FormControl fullWidth>
                        <StyledTextField
                        required
                        label="Todo Title"
                        variant="filled"
                        name='title'
                        defaultValue={todo.title}
                        />
                    </FormControl>
                    <FormControl fullWidth>
                        <TextField 
                        multiline
                        variant='filled'
                        label = "Todo description"
                        rows={5}
                        name = "description"
                        style={{
                            fontSize: "20px",
                            fontFamily : "arial",
                        }}
                        defaultValue={todo.body}
                        />
                    </FormControl>
                    <Stack spacing={2} direction="row-reverse">
                        <Button type='submit' variant='contained' color='secondary' size='medium'>
                            Save
                        </Button>
                        <Button variant='contained' color='error' size='medium' onClick={handleModalClose}>
                            Cancel
                        </Button>
                        <Tooltip title="Delete Todo" placement='left'>
                            <IconButton onClick={handleDeleteTodo}>
                                <Delete color="error"/>
                            </IconButton>
                        </Tooltip>
                    </Stack>
                </Stack>
            </form>
            
        </Box>
    </Modal>
  )
}

export default UpdateTodoModal;