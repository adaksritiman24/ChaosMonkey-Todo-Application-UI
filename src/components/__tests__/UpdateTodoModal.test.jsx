import TodoContext from "../../contexts/TodoContext";
import { SCOPE } from "../../constants/constants";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import UpdateTodoModal from "../UpdateTodoModal";
import { ACTIONS } from "../../constants/actions";
import { deleteTodo } from "../../utils/deleteTodo";

const dummyTodos = [
    {
        id : 1,
        title : "Dummy Title",
        body : "Dummy Body",
        completed : 0
    },
    {
        id : 2,
        title : "Dummy Title 2",
        body : "Dummy Body 2",
        completed : 1
    },
]


const dummyAppState = {
    todos : dummyTodos,
    scope : SCOPE.ALL,
    isUpdateTodoModalOpen : true
};
const dispatch = jest.fn();

const WithContext = (Application, modifiedState={})=> {
    const appState = {...dummyAppState, ...modifiedState}
    
    const todoContextValue = {appState, dispatch};
    return <TodoContext.Provider value={todoContextValue}>
            <Application/>
        </TodoContext.Provider>
}

jest.mock("../../utils/deleteTodo", ()=>{
    const actuals = jest.requireActual("../../utils/deleteTodo");
    return {
        ...actuals,
        __esModule : true,
        deleteTodo : jest.fn().mockImplementation(()=>Promise.resolve(1)),
    }
})

describe("UpdateTodoModal", ()=>{
    it("Should return empty react fragment when updatingtodo is undefined", ()=>{
        render(WithContext(UpdateTodoModal));

        expect(screen.queryByTestId("update-todo-modal")).not.toBeInTheDocument();
    })

    it("Should be in document and should match snapshot", ()=>{
        const {asFragment} = render(WithContext(UpdateTodoModal,{updatingTodo : dummyTodos[0]}));

        expect(screen.queryByTestId("update-todo-modal")).toBeInTheDocument();
        expect(asFragment).toMatchSnapshot();
    })

    it("should dispatch actions when modal is closed", ()=> {

        render(WithContext(UpdateTodoModal, {updatingTodo : dummyTodos[0]}));
        const cancelButtom = screen.getAllByRole("button")[1];

        fireEvent.click(cancelButtom);

        expect(dispatch).toBeCalledTimes(2);
        expect(dispatch).toBeCalledWith({
            type : ACTIONS.SET_IS_UPDATE_TODO_MODAL_OPEN,
            payload : false,
        });

        expect(dispatch).toBeCalledWith({
            type : ACTIONS.SET_UPDATING_TODO,
            payload : undefined,
        })
    })
    

    it("Should dispatch action when delete button is clicked", async()=>{
        render(WithContext(UpdateTodoModal, {updatingTodo : dummyTodos[0]}));
        const deleteButton = screen.getAllByRole("button")[2];

        fireEvent.click(deleteButton);

        await waitFor(()=>{
            expect(dispatch).toBeCalledWith({
                type : ACTIONS.SET_TODOS,
                payload : [dummyTodos[1]]
            })
        })

    })
})