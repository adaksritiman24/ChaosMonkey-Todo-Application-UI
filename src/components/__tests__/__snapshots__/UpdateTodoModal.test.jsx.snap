// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateTodoModal Should be in document and should match snapshot 1`] = `[Function]`;
